### PR TITLE
[Enhancement]  Add Gas Price Oracle User Configuration

### PR DESCRIPTION
**Pull request by Aishlia**
## Issue
#4357 Add feature to enable users to specify gas price parameters (blocks, percentile, defaultPriceGwei, maxPriceGwei, and numberTxsnSampled) in config and as flags. 

### Unit Test Coverage

Before:
```
ok      github.com/harmony-one/harmony/cmd/harmony    0.756s    coverage: 44.8% of statements
```

After:
```
PASS
ok  	github.com/harmony-one/harmony/cmd/harmony	0.568s
```

### Test/Run Logs
```
julianai@Harmonys-MacBook-Air harmony % go test ./cmd/harmony
# gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:23:13: warning: unused function '_duk_debugger_attach' [-Wunused-function]
ok  	github.com/harmony-one/harmony/cmd/harmony	0.495s
julianai@Harmonys-MacBook-Air harmony % go test -run TestHarmonyFlags
# gopkg.in/olebedev/go-duktape.v3
In file included from _cgo_export.c:4:
debugger.go:23:13: warning: unused function '_duk_debugger_attach' [-Wunused-function]
PASS
ok  	github.com/harmony-one/harmony/cmd/harmony	0.497s
```

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

2. **Describe the migration plan.**. For each flag epoch, describe what changes take place at the flag epoch, the anticipated interactions between upgraded/non-upgraded nodes, and any special operational considerations for the migration.

3. **Describe how the plan was tested.**

4. **How much minimum baking period after the last flag epoch should we allow on Pangaea before promotion onto mainnet?**

5. **What are the planned flag epoch numbers and their ETAs on Pangaea?**

6. **What are the planned flag epoch numbers and their ETAs on mainnet?**

    Note that this must be enough to cover baking period on Pangaea.

7. **What should node operators know about this planned change?**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

9. **Does the existing `node.sh` continue to work with this change?**

10. **What should node operators know about this change?**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**

    **NO**

### CODE REVIEW

File: rosetta/infra/harmony-pstn.conf [hunk 1]
```diff
@@ -120,3 +120,10 @@ Version = "2.5.11"
   Enabled = true
   IP = "0.0.0.0"
   Port = 9800
+
+[GasPrice]
+  Blocks = 20
+	Percentile = 80
+	DefaultPriceGwei = 9
+	MaxPriceGwei = 12
+  NumberTxsSampled = 3
```
Comment (MaxMustermann2) [inline rosetta/infra/harmony-pstn.conf:129]: Indent

File: cmd/harmony/main.go [hunk 2]
```diff
@@ -803,6 +803,7 @@ func setupConsensusAndNode(hc harmonyconfig.HarmonyConfig, nodeConfig *nodeconfi
 		Uint64("viewID", viewID).
 		Msg("Init Blockchain")
 
+	currentNode.HarmonyConfig.GasPrice = hc.GasPrice
```
Comment (MaxMustermann2) [inline cmd/harmony/main.go:806]: Why is this needed? 
Consider changing the name of this variable to `GasPriceOracleConfig`

File: cmd/harmony/flags.go [hunk 3]
```diff
@@ -1828,3 +1836,49 @@ func applyShardDataFlags(cmd *cobra.Command, cfg *harmonyconfig.HarmonyConfig) {
 		cfg.ShardData.CacheSize = cli.GetIntFlagValue(cmd, cacheSizeFlag)
 	}
 }
+
+var (
+	gasPriceBlocksFlag = cli.IntFlag{
+		Name:     "gasPrice.blocks",
```
Comment (MaxMustermann2) [inline cmd/harmony/flags.go:1842]: All of these need to be renamed similar to `gpo.blocks` per #4357

**Review by MaxMustermann2 — CHANGES_REQUESTED** [review]
Please update test `TestHarmonyFlags` as well.

File: rosetta/infra/harmony-mainnet.conf [hunk 4]
```diff
@@ -120,3 +120,10 @@ Version = "2.5.11"
   Enabled = true
   IP = "0.0.0.0"
   Port = 9800
+
+[GasPrice]
+  Blocks = 20
+	Percentile = 80
+	DefaultPriceGwei = 9
+	MaxPriceGwei = 12
+  NumberTxsSampled = 3
```
Comment (ONECasey) [inline rosetta/infra/harmony-mainnet.conf:129]: Same here

File: rosetta/infra/harmony-pstn.conf (hunk 1 above)
Comment (Aishlia) [reply rosetta/infra/harmony-pstn.conf:129]: Fixed. Not sure why tabs weren't working so I switched it to double space.

File: rosetta/infra/harmony-mainnet.conf (hunk 4 above)
Comment (Aishlia) [reply rosetta/infra/harmony-mainnet.conf:129]: Fixed.

File: cmd/harmony/main.go (hunk 2 above)
Comment (Aishlia) [reply cmd/harmony/main.go]: Added comment explaining this. Also changed the config from GasPriceConfig to GasPriceOracleConfig.

File: cmd/harmony/flags.go (hunk 3 above)
Comment (Aishlia) [reply cmd/harmony/flags.go]: Changed to 
```
--gpo.blocks
--gpo.percentile
--gpo.defaultprice
--gpo.maxprice
--gpo.txssampled 
```

File: cmd/harmony/flags.go [hunk 5]
```diff
@@ -1828,3 +1837,49 @@ func applyShardDataFlags(cmd *cobra.Command, cfg *harmonyconfig.HarmonyConfig) {
...
+		DefValue: defaultConfig.GasPriceOracle.Percentile,
+	}
+	gasPriceOracleDefaultPriceGweiFlag = cli.IntFlag{
+		Name:     "gpo.defaultprice",
+		Usage:    "Default gas price in gwei used when data is not available (default: 100)",
+		DefValue: defaultConfig.GasPriceOracle.DefaultPriceGwei,
+	}
+	gasPriceOracleMaxPriceGweiFlag = cli.IntFlag{
+		Name:     "gpo.maxprice",
+		Usage:    "Maximum possible gas price in gwei returned by eth_gasPrice (default: 12)",
```
Comment (MaxMustermann2) [inline cmd/harmony/flags.go:1859]: Why is max gas price (12) < default gas price (100) ?

File: cmd/harmony/flags_test.go [hunk 6]
```diff
@@ -159,6 +161,13 @@ func TestHarmonyFlags(t *testing.T) {
 					CacheTime:       10,
 					CacheSize:       512,
 				},
+				GasPriceOracle: harmonyconfig.GasPriceOracleConfig{
+					Blocks:           20,
+					Percentile:       80,
+					DefaultPriceGwei: 100,
+					MaxPriceGwei:     12,
+					NumberTxsSampled: 3,
```
Comment (MaxMustermann2) [inline cmd/harmony/flags_test.go:169]: Refer to the values here instead of hard coding them.

File: cmd/harmony/main.go [hunk 7]
```diff
@@ -803,6 +803,9 @@ func setupConsensusAndNode(hc harmonyconfig.HarmonyConfig, nodeConfig *nodeconfi
 		Uint64("viewID", viewID).
 		Msg("Init Blockchain")
 
+	// Assign GasPriceOracle parameters to the current node
+	currentNode.HarmonyConfig.GasPriceOracle = hc.GasPriceOracle
```
Comment (MaxMustermann2) [inline cmd/harmony/main.go:807]: I still don't see why this is needed. The `currentNode` stores within itself the entire config (including the `GasPriceOracle`) per below.
https://github.com/harmony-one/harmony/blob/b2e50e694ca7b7cd2adde45724711f9eafff049b/node/node.go#L1043

File: go.mod [hunk 8]
```diff
@@ -71,6 +71,7 @@ require (
 	github.com/c2h5oh/datasize v0.0.0-20220606134207-859f65c6625b
 	github.com/ledgerwatch/erigon-lib v0.0.0-20221218022306-0f8fdd40c2db
 	github.com/ledgerwatch/log/v3 v3.6.0
+	github.com/libp2p/go-libp2p-core v0.20.1
 )
```
Comment (MaxMustermann2) [inline go.mod:75]: It's preferred not to commit `go.mod` and `go.sum` for minor changes.

File: hmy/gasprice.go [hunk 9]
```diff
@@ -62,36 +61,88 @@ type Oracle struct {
...
+			Msg(
+				fmt.Sprint(
+					"Sanitizing invalid # txs sampled",
+					"provided",
+					params.NumberTxsSampled,
+					"updated",
+					DefaultGasPriceOracleConfig.NumberTxsSampled,
+				),
+			)
 	}
```
Comment (MaxMustermann2) [inline hmy/gasprice.go:138]: Add a sanity check for `maxPrice` < `defaultPrice`

File: hmy/staking.go [hunk 10]
```diff
@@ -387,7 +387,7 @@ func (hmy *Harmony) GetValidatorInformation(
 	// b.apiCache.Forget(prevKey)
 
 	// calculate last APRHistoryLength epochs for averaging APR
-	// epochFrom := bc.GasPriceConfig().StakingEpoch
+	// epochFrom := bc.GasPriceOracleConfig().StakingEpoch
```
Comment (MaxMustermann2) [inline hmy/staking.go:390]: Undo this change.

File: hmy/hmy.go [hunk 11]
```diff
@@ -157,11 +159,8 @@ func New(
 	}
 
 	// Setup gas price oracle
-	gpoParams := GasPriceConfig{
-		Blocks:     20,                // take all eligible txs past 20 blocks and sort them
-		Percentile: 80,                // get the 80th percentile when sorted in an ascending manner
-		Default:    big.NewInt(100e9), // minimum of 100 gwei
-	}
+	gpoParams := nodeAPI.GetHarmonyConfig().GasPriceOracle
+
```
Comment (MaxMustermann2) [inline hmy/hmy.go:163]: nit: Remove this blank line.

**Review by MaxMustermann2 — CHANGES_REQUESTED** [review]

**Comment by ONECasey** [comment]
A few merge conflicts after a rebase.

**Comment by ONECasey** [comment]
New dev. Needs mentorship. Closed
